include/utilities/parallel_utils.h

```diff
@@ -127,6 +131,297 @@ namespace parallel {
+                int matches_host_at = -1;
+
+                for (int known_rank = 0; known_rank < recv_rank; ++known_rank) {
+                    if (std::strcmp(hostnames[known_rank], hostnames[recv_rank]) == 0) {
+                        matches_host_at = known_rank;
+                        break;
+                    }
+                }
+                if (matches_host_at == -1) {
+                    // Assign new host id, then increment what the next id will be
+                    host_array[recv_rank] = next_host_id++;
+                }
+                else {
+                    host_array[recv_rank] = host_array[matches_host_at];
+                }
+            }
+        }
+        // Now, broadcast the results out
+        MPI_Bcast(host_array, mpi_num_procs, MPI_INT, 0, MPI_COMM_WORLD);
+    }
+
+    /**
+     * Send the contents of a text file to another MPI rank.
+     *
+     * Note that the file is read with ``fgets``.
+     *
+     * @param fileName The text file to read and send its contents.
+     * @param mpi_rank The current MPI rank.
+     * @param destRank The MPI rank to which the file data should be sent.
+     * @return Whether sending was successful.
+     */
+    bool mpi_send_text_file(const char *fileName, const int mpi_rank, const int destRank) {
+        int bufSize = 4096;
+        char buf[bufSize];
+        int code;
+        // How much has been transferred so far
+        int totalNumTransferred = 0;
+
+        FILE *file = fopen(fileName, "r");
+
+        // Transmit error code instead of expected size and return false if file can't be opened
+        if (file == NULL) {
+            // TODO: output error message
+            code = -1;
+            MPI_Send(&code, 1, MPI_INT, destRank, 0, MPI_COMM_WORLD);
+            return false;
+        }
+
+        // Send expected size to start
+        MPI_Send(&bufSize, 1, MPI_INT, destRank, 0, MPI_COMM_WORLD);
+
+        // Then get back expected size to infer other side is good to go
+        MPI_Recv(&code, 1, MPI_INT, destRank, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+        if (code != bufSize) {
+            // TODO: output error message
+            fclose(file);
+            return false;
+        }
+
+        // Then while there is more of the file to read and send, read the next batch and ...
+        while (fgets(buf, bufSize, file) != NULL) {
+            // First send this batch
+            MPI_Send(buf, bufSize, MPI_CHAR, destRank, 0, MPI_COMM_WORLD);
+            // Then get back a code, which will be -1 if bad and need to exit and otherwise good
+            MPI_Recv(&code, 1, MPI_INT, destRank, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+            if (code == -1) {
+                // TODO: output error message
+                fclose(file);
+                return false;
+            }
+        }
+        // Once there is no more file to read and send, send an empty string
+        MPI_Send("", bufSize, MPI_CHAR, destRank, 0, MPI_COMM_WORLD);
```

Comment:
What happens when the text file has a blank line in it?  Will the receiver continue as expected to receive the next line?  It doesn't look like it, the receiver will close the writing file and break the receiving loop.  But the sending is going continue to the next line (since it got a code of 0...the str length of the offending send), and it will hang waiting for the receiver to receive.   Or does `fgets` read newlines and consider that a line of length 1?  I guess the `while` will catch that case, since `fgets` doesn't read a single new line line, the pointer will be null, but then you might miss some of the file due to early stop.  So this function can only `send_text_file_up_to_first_blank_line` I think.

This might be a good place to use an MPI_TAG to indicate the difference between a data message and protocol message:

```c++
        while (fgets(buf, bufSize, file) != NULL) {
            // First send this batch
            MPI_Send(buf, bufSize, MPI_CHAR, destRank, DATA_TAG, MPI_COMM_WORLD);
            // Then get back a code, which will be -1 if bad and need to exit and otherwise good
            MPI_Recv(&code, 1, MPI_INT, destRank, DATA_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
            if (code == -1) {
                // TODO: output error message
                fclose(file);
                return false;
            }
        }
        // Once there is no more file to read and send, send an empty string
        MPI_Send("", bufSize, MPI_CHAR, destRank, PROTOCOL_TAG, MPI_COMM_WORLD);
```

Then in receive, you can match the `PROTOCOL_TAG` in the appropriate protocol messages.